src/controllers/user/me.rs

```diff
@@ -212,3 +227,38 @@ pub fn regenerate_token_and_send(req: &mut dyn Request) -> CargoResult<Response>
     }
     Ok(req.json(&R { ok: true }))
 }
+
+/// Handles `PUT /me/email_notifications` route
+pub fn update_email_notifications(req: &mut dyn Request) -> CargoResult<Response> {
+    use self::crate_owners::dsl::*;
+    use diesel::update;
+
+    #[derive(Deserialize)]
+    struct CrateEmailNotifications {
+        id: i32,
+        email_notifications: bool,
+    }
+
+    let mut body = String::new();
+    req.body().read_to_string(&mut body)?;
+    let updates: Vec<CrateEmailNotifications> =
+        serde_json::from_str(&body).map_err(|_| human("invalid json request"))?;
+
+    let user = req.user()?;
+    let conn = req.db_conn()?;
+
+    for to_update in updates {
+        update(CrateOwner::belonging_to(user))
```

Comment:
So this query will need a filter on `owner_kind` too, like the query above. Sorry again for forgetting to include this in the instructions!

I also realize I didn't communicate clearly about what I was thinking for the updates here-- I suggested updating the records unconditionally (rather than checking to see if they needed to be updated as you're doing here with the `filter(email_notifications.ne(to_update.email_notifications))`) so that instead of doing one update statement per crate owned, we can do one `INSERT... ON CONFLICT DO UPDATE` (aka "upsert") statement including all the crate ownership records for this user. There are [some people who own a lot of crates](https://crates.io/users/alexcrichton), so while in most cases doing an UPDATE per record would be fine, the edge cases worry me!

[Here's an example of doing an upsert](https://github.com/rust-lang/crates.io/blob/2efe33df7183c2822e8b1b53e1090a510b6b6fce/src/boot/categories.rs#L114-L123) with multiple records like needs to be done here, are you up for making that change? Please let me know if more details would be helpful!

Reply:
I pushed a commit to address this one. The one downside that I think the `updated_at` column would be updated regardless of whether it was changed or not.

Reply:
> The one downside that I think the `updated_at` column would be updated regardless of whether it was changed or not.

I don't think that's the case -- updating the `updated_at` column [is implemented as a `BEFORE UPDATE` per-row trigger](https://github.com/rust-lang/crates.io/blob/36cfbe31b3a81d3a80782299673e3925edde4758/migrations/20151126095136_dumped_migration_103/up.sql#L47-L49), and if I'm reading [the postgres docs](https://www.postgresql.org/docs/12/trigger-definition.html) correctly, these only apply to *affected* rows:

> With a per-row trigger, the trigger function is invoked once for each row that is affected by the statement that fired the trigger.

So inserting with a conflict but the same values that are already there wouldn't affect that row. I did some testing locally and the behavior I saw was consistent with my understanding of the docs.

HOWEVER, this does raise the issue of whether we want the `updated_at` column to be changed when someone DOES update this preference... and now I'm conflicted (pun not intended)!

- The updated_at column on crate_owners is currently public information [at least through the database dumps](https://github.com/rust-lang/crates.io/blob/36cfbe31b3a81d3a80782299673e3925edde4758/src/tasks/dump_db/dump-db.toml#L69) when [the record isn't `deleted`](https://github.com/rust-lang/crates.io/blob/36cfbe31b3a81d3a80782299673e3925edde4758/src/tasks/dump_db/dump-db.toml#L62). There's a possibility one could figure out if you updated your email notification settings if we changed the updated_at. While probably unlikely and probably not valuable, I'd like to not leak info if we can help it.
- The main purpose of that `updated_at` column is to help in auditing when we're trying to figure out why something unexpected happened with who owns a crate. This is fairly unhelpful, though, as we aren't recording *who* updated the record. [We're working on adding more auditing](https://github.com/rust-lang/crates.io/issues/1548) but that's not done yet. Once that auditing is implemented, I'm not sure that we care about this column at all, but I need to figure that out some more.
- Maybe this is another sign that this table is the wrong place for storing this preference...... but I'm not sure where the right place would be. A new table? Just for the "off" settings because we're defaulting to "on"? Do you have any thoughts? I'm going to bounce some ideas around with some folks and I'll update here with what I'm thinking in the next few days.

Reply:
The ON CONFLICT logic is helpful to know. Thanks for researching that.

I don't personally like the idea of a "don't send me notifications for this crate" table. How it is seems okay, except for the new column not making sense for teams. I'll be curious to know what the general consensus is.

If we stick with the status quo for this PR, would it make sense to remove the updated_at column from the DB dump?

Reply:
> I don't personally like the idea of a "don't send me notifications for this crate" table. How it is seems okay, except for the new column not making sense for teams. I'll be curious to know what the general consensus is.

I chatted with @jtgeibel at a meetup last night, and he didn't like that this column didn't make sense for teams. But that's kind of an unrelated problem-- the solution is to split this table up into teams only and users only, and that doesn't have anything to do with this change.

I think removing the `updated_at` column from the database dump makes sense, and we'll get to see if anyone has complaints about us removing columns! (We have said the database dumps are experimental, so they can't be too mad). Let's go with that. Thanks!

Reply:
I made `updated_at` private in the latest commit.